dynomite/src/lib.rs

```diff
@@ -31,7 +31,30 @@
 //!   user: Uuid,
 //!   #[dynomite(sort_key)]
 //!   order_id: Uuid,
-//!   color: Option<String>
+//!   color: Option<String>,
+//! }
+//!
+//! #[derive(Item)]
+//! struct ShoppingCart {
+//!     #[dynomite(partition_key)]
+//!     id: Uuid,
+//!     // A separate struct to store data without any id
+//!     #[dynomite(flatten)]
```

Comment:
Can we separate these examples to distinguish the use case where you would find flatten most useful?

The current api aims to make a mapping between structs and dynamo item schema which is intuative enough that you can guess what the struct will look like when looking at the dynamo item and visa versa. I'd like to still promote that as the default path until there's a need or special case where you mind need flattening.

Let's make a second separate example describing where in when you might need this.

Reply:
Flattening is required to support flat fat enums:
```rust
#[derive(Item)] 
struct MyItem {
    #[dynomite(partition_key)]
    id: String,
    #[dynomite(flatten)]
    my_enum: Foo,
}

#[dynomite(tag = "kind")]
enum Foo {
    Bar(Bar),
}

#[derive(Attributes)]
struct Bar {
    bar: u32
}
```
Without that `flatten` attribute the item would be serialized as:
```jsonc
{
    "id": "1234",
    "my_enum": {
         "kind": "Bar", // can't use neither kind, nor bar as a sort key or GSI primary key
         "bar": 42,
    }
}
```
which won't work if you want to make `bar` or `kind` fields a partition or sort key.
With flatten this becomes possible:
```jsonc
{
     "id": "1234"
     "kind": "Bar",
     "bar": 42,
}
```
I've started with `flatten`, so I can't give an example with a fat enum right now in this PR.
Otherwise `flatten` is useful to extract repetitive data, maybe even construct the items itself from disjoint projections,
or just have the `id`-less struct that contains only datums (this is useful when `id` is not yet known):
```rust
#[derive(Item)]
struct User {
    #[dynomite(partition_key)]
    id: Uuid,
    #[dynomite(flatten)]
    profile: UserProfile,
}

#[derive(Attributes)]
struct UserProfile {
    name: String,
    avatar: Url,
}

// Accepts data, no id field is required to be filled by the caller
async fn create_user(data: UserProfile) -> Result<User> {
     let user = User {
          id: Uuid::new_v4(),
          profile: data,
     };
     db.write_item({ user.clone()... }).await?;
     user
}
```
Though for the last point we may just add yet another proc macro to generate struct without the partition key field and update changeset structs with all fields as `Option<T>`.

In this example here I have shown extracting the data from `id` into a separate struct because I cannot show an example with a flat fat enum yet...